### PR TITLE
Update jane dates

### PR DESCRIPTION
**Pull request by KeeyanGhoreshi**
This makes some minor changes to Jane Bradshaw's record which puts the consult request for the breast cancer condition ahead of the diagnosis chronologically, so that the patient updater doesn't put the diagnosis date into the future.  Her procedure requests also now fall before the current encounter, which is id `33`.  A duplicate entry with id `33` was renamed to have id `50`, which was the next in incremental order. 

There's some potential changes that I was unsure about.  I avoided making any of these changes, but it seems to me that some resources referencing the breast cancer condition should occur after the diagnosis.  For example, many of the observations have a `RelevantTime` before the diagnosis, but reference the breast cancer condition as a `SpecificFocusOfFinding`.   The actual chronology is of course beyond my knowledge, but I am unsure whether certain events are supposed to be before or after the diagnosis.  It seems a bit strange to be getting information on tumor size and stage on dates well before the diagnosis of cancer.  

This PR also gives the patient date updater two new flags for outputting the date values.  Using `-o` or `--output` will output all found dates and the entry type they belong to, as well as their immediate context.  Using `-O` or `--ordered-output` will output the dates ordered chronologically.

You don't have to put in an encounter `id` if you just want to output the list of dates.  Also, an important note, if you run the patient updater through the script using `npm`, you have to pass flags in using a `--` to separate them, as discussed [here](https://github.com/npm/npm/pull/5518).  So to run the patient updater and output the dates only, you'd run `npm run patient-date-update patient.json -- -o`

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [x] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [x] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [ ] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [x] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [x] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [x] Document any new APIs/extension points
- [x] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- [x] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by danlee1025** [comment]
This branch needs to be rebased.

**Comment by danlee1025** [comment]
This looks good to me.  I like the new functionality of outputting all of the dates and the dates in the UI now make sense(no events are occurring in the future when the date updater is run).

@gregquinn2001 Did you want to comment on the `RelevantTime` of many observations being before the diagnosis date?  I don't think these dates are displayed anywhere on `/ccp` so the UI looks fine but did we want to fix these dates if they are indeed in the wrong order?

**Comment by gregquinn2001** [comment]
I will take a look! Thanks Keeyan and Dan!

**Comment by gregquinn2001** [comment]
[GQ1] Please update the "Patient Date Update" section of README to incorporate new functionality. Please also fix the example given to use a valid file name (I know that wasn't from your change)

**Comment by gregquinn2001** [comment]
[GQ2] Can we get rid of the hard to read blue color and just leave the colors in the console alone? It's really hard to read on my console at least.

**Comment by gregquinn2001** [comment]
[GQ3] The chronological output needs to use the key date from the entry to be useful. creation date and update date are not the key clinical date. Perhaps use last modified if only those 2 dates exist. If it has another date though, use that. If it has multiple other dates, I guess use the first one????

**Comment by gregquinn2001** [comment]
I ran the following:

yarn patient-date-update ./src/dataaccess/BreastMainTreatmentDiabetesHypertensionJaneV05.json -- -o

and it modified the json file

**Comment by KeeyanGhoreshi** [comment]
The newest commit addresses [GQ1] and [GQ2].  For [GQ3], I have the chronological date output filter out metadata dates if there is a non-metadata value, but I'm still unclear on what to do if there's multiple dates that aren't metadata.  I'd think that you'd want to see them all, but I also could see only wanting the most recent one to show up in the list.  

I couldn't recreate [GQ4], running the same exact command I do not get a change in the JSON, the backup still gets made but no changes to the json file that gets passed in.  I should also note that yarn doesn't require the `--` like npm does.  As I wrote in the `readme`, you can just run `yarn patient-date-update ./src/dataaccess/BreastMainTreatmentDiabetesHypertensionJaneV05.json -o`.  I'm running yarn v1.15.2.

**Comment by gregquinn2001** [comment]
Thanks Keeyan. I'll take a look. I had specified the output but somehow it looks like it's gone in GQ4. sorry about that. You can see the incomplete sentence "The output should be something like ." I was trying to say let's have a yarn script entry that does the correct command line options and just accepts the json file name and call it patient-date-list. It can provide the -o option. The output should always be chronological. I'm not sure what the use case is for non-chronological. It would output a line per entry with clinical date as follows:

<entry type> <entry id> <1st clinical date attribute name>:<1st clinical date value> <2nd clinical date attribute name>:<2nd clinical date value> ... LastUpdated: <last updated value> Authored: <Authored value>

If an entry doesn't have any dates other than created and updated let's not show it (e.g., Patient).

**Comment by gregquinn2001** [comment]
For GQ4, you did not even mention yarn in your pull request description and you added the readme content based on GQ1 so I didn't have it available!

I still have the same problem but I have yarn v 1.7.0 so I'll try to update yarn and see if it goes away

**Comment by gregquinn2001** [comment]
For GQ4, I updated yarn. It still happened:
image
image

**Comment by gregquinn2001** [comment]
Is it possible you are still writing the output file but it's the same as the input for you? In my case it may have changed the line endings so git diff shows a change.

**Comment by KeeyanGhoreshi** [comment]
Yes that's probably it.  The file still gets written out.  I'll change it to not write out to file if no encounter is provided.

**Comment by KeeyanGhoreshi** [comment]
Also sorry for not mentioning yarn, I hadn't even tested with it until you mentioned it in your comments.  Didn't mean to imply that the instructions were available, just noting that I included those instructions in the readme.

**Comment by KeeyanGhoreshi** [comment]
Latest commit should address [GQ4].  You can now do `yarn patient-date-list ./src/dataaccess/BreastMainTreatmentDiabetesHypertensionJaneV05.json` and the updater will no longer write out the file unless it is given an encounter.

**Comment by danlee1025** [comment]
[DL1]  The backup file is still being written out even when an encounter isn't given.  We probably don't need this step if the original file isn't going to be overwritten.

File: src/patient-date-updater/app.js [hunk 1]
```diff
@@ -20,55 +21,94 @@ if (typeof input === 'undefined') {
...
+                date.add(deltaDuration.asDays(), 'd');
+                flattenedEntry[key] = date.format('D MMM YYYY HH:mm ZZ');
+                change = true;
+            }
+            entry.date = date;
+        }
+        if (orderedOutput && isDate) {
+            entry.key = key;
+            entry.value = value;
+            if(entries[entry.entryId]){
```
Comment (danlee1025) [inline src/patient-date-updater/app.js]: [DL2] Fix spacing from lines 105-109.

File: src/patient-date-updater/app.js [hunk 2]
```diff
@@ -77,9 +117,61 @@ patientEntries.forEach((entry, i) => {
...
+        console.log('DONE');
+    });
+}
+
+
+function log(key, value) {
+    console.log(`\t\t${key}: ${value}`);
+}
+
+function flattenOrderedOutput(entries) {
```
Comment (danlee1025) [inline src/patient-date-updater/app.js]: [DL3] Fix spacing in `flattenOrderedOutput` function.

**Comment by danlee1025** [comment]
[DL4]  Add `patient-date-list` command to README.

**Comment by gregquinn2001** [comment]
I updated the patient-date-list capability to be usable for desired use case of validating the patient record's dates. A lot of dates should be fixed. Anything that references 8 should have a date later than entry 8 onset.  I did not address them as I said we didn't need to unless they affected the UI.

**Pull request merged**